refactor: concurrency safe refactor

tynany opened the pull request:
- Cleaner exporter structure
- Fix concurrency safety issues mentioned in https://github.com/tynany/frr_exporter/issues/47

bdrung commented on collector/bgp.go:
```diff
@@ -8,155 +8,88 @@ import (
-		ch <- desc
-	}
+type bgpCollector struct {
+	logger log.Logger
```
(1) The global variables `bgpDesc` and `bgpL2vpnDesc` should be moved here. Then (2)

tynany replied:
Done. I have applied this to all other collectors as well.

bdrung commented on collector/bgp.go:
```diff
@@ -8,155 +8,88 @@ import (
-	collectBGP(ch, "ipv4")
+// NewBGPCollector collects BGP metrics, implemented as per the Collector interface.
+func NewBGPCollector(logger log.Logger) (Collector, error) {
+	return &bgpCollector{logger: logger}, nil
```
(2) The whole code from `setBgpDesc` should be moved here, since it's enough to execute it once. Then `bgpDescMtx` is not needed any more (since you only create the data here and later only read it).

tynany replied:
Moving all code from `setBgpDesc` breaks the tests. I can't think of a cleaner way than leaving `setBgpDesc` in place. Thoughts?

`bgpDescMtx` has now been removed, though, thanks to (1).

bdrung replied:
Your implementation to have `getBGPDesc` and call it in `NewBGPCollector` and `NewBGP6Collector` looks good to me.

bdrung commented on collector/bgp.go:
```diff
@@ -8,155 +8,88 @@ import (
-// CollectTotalErrors returns total errors.
-func (*BGPCollector) CollectTotalErrors() float64 {
-	return totalBGPErrors
+type bgp6Collector struct {
```
You can merge the `bgp6Collector` and `bgpCollector` structs by adding a variable `AFI string` to it. Then you can
* drop the `Update` function for `bgp6Collector`
* move collectBGP into the `Update` function

tynany replied:
Nice! Done.

bdrung reviewed (changes requested):
Running the test cases fail:
```
$ go test ./... -v
?       github.com/tynany/frr_exporter  [no test files]
=== RUN   TestProcessBFDPeers
--- PASS: TestProcessBFDPeers (0.00s)
=== RUN   TestProcessBGPSummary
--- FAIL: TestProcessBGPSummary (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x60 pc=0x6ee41d]

goroutine 35 [running]:
testing.tRunner.func1.2({0x76fba0, 0xadaee0})
        /usr/lib/go-1.17/src/testing/testing.go:1209 +0x24e
testing.tRunner.func1()
        /usr/lib/go-1.17/src/testing/testing.go:1212 +0x218
panic({0x76fba0, 0xadaee0})
        /usr/lib/go-1.17/src/runtime/panic.go:1038 +0x215
github.com/prometheus/client_golang/prometheus.NewConstMetric(0xaeb6c0, 0xc00021aec0, 0x0, {0xc00021aec0, 0xc0000d5ae8, 0x40eec7})
        /home/bdrung/go/pkg/mod/github.com/prometheus/client_golang@v1.11.0/prometheus/value.go:88 +0x1d
github.com/prometheus/client_golang/prometheus.MustNewConstMetric(...)
        /home/bdrung/go/pkg/mod/github.com/prometheus/client_golang@v1.11.0/prometheus/value.go:105
github.com/tynany/frr_exporter/collector.newGauge(0x7d2e54, 0x548f65, 0x7d3385, {0xc00021aec0, 0xc000220190, 0xc00026c028})
        /home/bdrung/git/backports/prometheus-frr-exporter/collector/collector.go:140 +0x31
github.com/tynany/frr_exporter/collector.processBGPSummary(0xc00021c720, {0xaa3120, 0x804, 0x804}, {0x7d2064, 0x4}, {0x7d2e54, 0x7}, {0x0, 0x0})
        /home/bdrung/git/backports/prometheus-frr-exporter/collector/bgp.go:227 +0x465
github.com/tynany/frr_exporter/collector.TestProcessBGPSummary(0xc000210b60)
        /home/bdrung/git/backports/prometheus-frr-exporter/collector/bgp_test.go:387 +0x6c
testing.tRunner(0xc000210b60, 0x7f3648)
        /usr/lib/go-1.17/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
        /usr/lib/go-1.17/src/testing/testing.go:1306 +0x35a
FAIL    github.com/tynany/frr_exporter/collector        0.006s
FAIL
```

bdrung commented:
Side question (in case you want to change code, this should be done in a follow up commit): In `getBGPPeerDesc` why do you run the command `show run bgpd` and parse that output instead of calling `show bgp neighbors json` to get a JSON response which includes `nbrDesc`?

dswarbrick commented on collector/collector.go:
```diff
@@ -1,159 +1,131 @@
 // Collect implemented as per the prometheus.Collector interface.
-func (e *Exporters) Collect(ch chan<- prometheus.Metric) {
+func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+	frrTotalScrapeCountMtx.Lock()
 	frrTotalScrapeCount++
+	frrTotalScrapeCountMtx.Unlock()
```
Consider replacing this with a [prometheus.Counter](https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#Counter), which are concurrency-safe and don't require you to implement your own locking.

tynany replied:
Good call. Done.

dswarbrick commented on collector/pim.go:
```diff
@@ -5,93 +5,49 @@ import (
+	pimNeighborLabels = append(pimLabels, "neighbor", "iface")
+	pimDesc           = map[string]*prometheus.Desc{
+		"neighborCount": colPromDesc(pimSubsystem, "neighbors_count_total", "Number of neighbors detected", pimLabels),
+		"upTime":        colPromDesc(pimNeighborMetrixPrefix, "uptime_seconds", "How long has the peer ben up.", pimNeighborLabels),
```
Typo
```suggestion
		"upTime":        colPromDesc(pimNeighborMetrixPrefix, "uptime_seconds", "How long has the peer been up.", pimNeighborLabels),
```

tynany replied:
Fixed.

dswarbrick commented on collector/vrrp.go:
```diff
@@ -46,83 +59,26 @@ type VrrpInstanceStats struct {
 	if err != nil {
-		totalVRRPErrors++
-		vrrpErrors = append(vrrpErrors, fmt.Errorf("cannot get vrrp info: %s", err))
+		return fmt.Errorf("cannot get vrrp info: %s", err)
```
Consider wrapping errors using the `%w` verb where you include it in a `fmt.Errorf()`, e.g.
```suggestion
		return fmt.Errorf("cannot get vrrp info: %w", err)
```

tynany replied:
Fixed.

tynany commented:
> Running the test cases fail:
> 
> ```
> $ go test ./... -v
> ?       github.com/tynany/frr_exporter  [no test files]
> === RUN   TestProcessBFDPeers
> --- PASS: TestProcessBFDPeers (0.00s)
> === RUN   TestProcessBGPSummary
> --- FAIL: TestProcessBGPSummary (0.00s)
> panic: runtime error: invalid memory address or nil pointer dereference [recovered]
>         panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x1 addr=0x60 pc=0x6ee41d]
> 
> goroutine 35 [running]:
> testing.tRunner.func1.2({0x76fba0, 0xadaee0})
>         /usr/lib/go-1.17/src/testing/testing.go:1209 +0x24e
> testing.tRunner.func1()
>         /usr/lib/go-1.17/src/testing/testing.go:1212 +0x218
> panic({0x76fba0, 0xadaee0})
>         /usr/lib/go-1.17/src/runtime/panic.go:1038 +0x215
> github.com/prometheus/client_golang/prometheus.NewConstMetric(0xaeb6c0, 0xc00021aec0, 0x0, {0xc00021aec0, 0xc0000d5ae8, 0x40eec7})
>         /home/bdrung/go/pkg/mod/github.com/prometheus/client_golang@v1.11.0/prometheus/value.go:88 +0x1d
> github.com/prometheus/client_golang/prometheus.MustNewConstMetric(...)
>         /home/bdrung/go/pkg/mod/github.com/prometheus/client_golang@v1.11.0/prometheus/value.go:105
> github.com/tynany/frr_exporter/collector.newGauge(0x7d2e54, 0x548f65, 0x7d3385, {0xc00021aec0, 0xc000220190, 0xc00026c028})
>         /home/bdrung/git/backports/prometheus-frr-exporter/collector/collector.go:140 +0x31
> github.com/tynany/frr_exporter/collector.processBGPSummary(0xc00021c720, {0xaa3120, 0x804, 0x804}, {0x7d2064, 0x4}, {0x7d2e54, 0x7}, {0x0, 0x0})
>         /home/bdrung/git/backports/prometheus-frr-exporter/collector/bgp.go:227 +0x465
> github.com/tynany/frr_exporter/collector.TestProcessBGPSummary(0xc000210b60)
>         /home/bdrung/git/backports/prometheus-frr-exporter/collector/bgp_test.go:387 +0x6c
> testing.tRunner(0xc000210b60, 0x7f3648)
>         /usr/lib/go-1.17/src/testing/testing.go:1259 +0x102
> created by testing.(*T).Run
>         /usr/lib/go-1.17/src/testing/testing.go:1306 +0x35a
> FAIL    github.com/tynany/frr_exporter/collector        0.006s
> FAIL
> ```

Fixed:

```
% go test
PASS
ok  	github.com/tynany/frr_exporter/collector	0.316s
```

tynany commented:
> 

That is a much better implementation, thanks! I have pushed this change to this branch/PR.

tynany commented:
@dswarbrick @bdrung thanks for reviewing! I have replied to all comments.

FYI, in my latest commit, I finally got around to implementing a change I've been meaning to do for a while -- interfacing with FRR via UNIX sockets instead of `vtysh`. No more forking cmds.

bdrung commented on collector/pim.go:
```diff
@@ -11,52 +11,47 @@ import (
+	neighborLabels := append(labels, "iface", "neighbor")
+	return map[string]*prometheus.Desc{
+		"neighborCount": colPromDesc(pimSubsystem, "neighbors_count_total", "Number of neighbors detected", labels),
+		"upTime":        colPromDesc(pimSubsystem+"_neighbor", "uptime_seconds", "How long has the peer been up.", neighborLabels),
```
Why not
```
colPromDesc(pimSubsystem, "neighbor_uptime_seconds",[...]
```

Besides that it looks inconsistent to use neighbor and neighbors (singular and plural).

tynany replied:
Agreed. Fixed.

bdrung commented on collector/bgp.go:
```diff
@@ -126,54 +149,7 @@ func processBgpL2vpnEvpnSummary(ch chan<- prometheus.Metric, jsonBGPL2vpnEvpnSum
-		"peerGroupCount":  colPromDesc(bgpSubsystem, "peer_groups_count_total", "Number of peer groups configured.", bgpLabels),
-		"peerGroupMemory": colPromDesc(bgpSubsystem, "peer_groups_memory_bytes", "Memory consumed by peer groups.", bgpLabels),
-
-		"msgRcvd":               colPromDesc(bgpPeerMetricPrefix, "message_received_total", "Number of received messages.", bgpPeerLabels),
```
Why not
```
colPromDesc(bgpSubsystem, "peer_message_received_total", [...]
```
?

tynany replied:
I am using another variable, `bgpPeerMetricPrefix`, for convenience as there are 7 peer metrics. Is there an advantage of doing:

```
colPromDesc(bgpSubsystem, "peer_message_received_total", [...]
colPromDesc(bgpSubsystem, "peer_message_sent_total", [...]
...
```

rather than:
```
bgpPeerMetricPrefix := "bgp_peer"
colPromDesc(bgpPeerMetricPrefix, "peer_message_received_total", [...]
colPromDesc(bgpPeerMetricPrefix, "peer_message_sent_total", [...]
...
?

bdrung replied:
It's purely for consistency, since you also have four metrics like `peerGroupCount` compared to these 7 peer metrics.

bdrung commented on collector/vrrp.go:
```diff
@@ -60,48 +48,57 @@ type VrrpInstanceStats struct {
+
+	return map[string]*prometheus.Desc{
+		"vrrpState":       colPromDesc(vrrpSubsystem, "state", "Status of the VRRP state machine.", stateLabels),
+		"adverTx":         colPromDesc(vrrpSubsystem, "adverTx_total", "Advertisements sent total.", labels),
```
adverTx sounds cryptic to me. "advertisements_send_total" and using "received" for Rx would be clearer. In case you change that, please put it into a separate commit.

tynany replied:
I agree. I will fix this in a subsequent PR.

bdrung commented:
Thanks for implementing and incorporating our feedback. Except from a few comments I am fine with this PR now.

The only thing I would have preferred is to have atomic commits (e.g. split "PR comment fixes" and "move to sockets" into two separate commits).

tynany commented:
> Thanks for implementing and incorporating our feedback. Except from a few comments I am fine with this PR now.
> 
> The only thing I would have preferred is to have atomic commits (e.g. split "PR comment fixes" and "move to sockets" into two separate commits).

Thanks for the quick review.

You are right, I shouldn't have added the "move to sockets" change to this PR. I have rebased this branch, removing the socket changes and will add them to another PR.
